src/hotspot/share/services/management.cpp

```diff
@@ -748,7 +750,7 @@ JVM_ENTRY(jobject, jmm_GetMemoryUsage(JNIEnv* env, jboolean heap))
 
     for (int i = 0; i < MemoryService::num_memory_pools(); i++) {
       MemoryPool* pool = MemoryService::get_memory_pool(i);
-      if (pool->is_non_heap()) {
+      if (pool->is_codeheap() || pool->is_metaspace()) {
```

Comment:
Our only special case is that all the memory reported by `CompressedKlassSpacePool` are already reported by `MetaspacePool`, so shouldn't we do this:

```
if (pool->is_non_heap() && !pool->is_compressed_klass_space()) {
  // skip CompressedKlassSpacePool since its memory is already reported by
  // MetaspacePool
```

Reply:
This is also acceptable, but if we add more memory pools, we might add more special pools to exclude their space from nonheap. Instead, we can specify exactly which pools we need to accumulate, which is what this change did.

Reply:
I think the problem is the definition of the pools. We seem to have nested pools but it is far from clear that this API/mechanism was designed/intended to support nested pools.

Reply:
In any case, it's unreasonable for getNonHeapMemoryUsage to repeatedly count a piece of memory, right? In the extreme case, we might add all nested pools of metaspace, and we will get Metaspace x2 + CodeCache when using getNonHeapMemoryUsage.

From the developer's point of view, MemoryMXBean.getNonHeapMemoryUsage is expected to obtain the size of non-heap area. Given that MemoryMXBean.getHeapMemoryUsage is clearly stated which is heap area:
```
     * Returns the current memory usage of the heap that
     * is used for object allocation.  The heap consists
     * of one or more memory pools.
```

I propose to revise the Java doc to describe the definition of non-heap area more precisely:
```
    /**
     * Returns the current memory usage of non-heap memory that
     * contains code cache and metaspace.
     * The non-heap memory consists of one or more memory pools.
```

Reply:
No, things like code-cache and metaspace do not get baked into the specification. "non-heap" is deliberately vague and open-ended.

I can agree the calculation may be wrong but we need to determine exactly why it is wrong. To me the pool definitions seem wrong - they shouldn't be nested, so then you couldn't double-count. But we really need the experts on this code to chime in.

Reply:
Sorry for chiming in a not very helpful way, but I am not sure what even the point is of this API. 

I'm seriously interested in who uses it, and for what. Calculating real memory usage is notoriously difficult. This one seems to be just an arbitrary selection of stuff. It leaves out C-heap usage (VM, outer JDK, and third party libs), native mappings, sometimes massive system library overhead, text segments, thread stacks... that it counts class space twice is probably its smallest problem...

Reply:
Beside my general points, I think this patch makes sense. I agree with @kelthuzadx that counting Classspace twice is wrong, whatever the intention of this API originally was.

Reply:
Yes counting it twice is wrong but this is the wrong fix. The pools are not defined correctly.

Reply:
> Sorry for chiming in in a not very helpful way, but I am not sure what even the point is of this API.
> 
> I'm seriously interested in who uses it, and for what. Calculating real memory usage is notoriously difficult. This one seems to be just an arbitrary selection of stuff. It leaves out C-heap usage (VM, outer JDK, and third party libs), native mappings, sometimes massive system library overhead, text segments, thread stacks... that it counts class space twice is probably its smallest problem...

AFAIK, some DevOps platforms monitor JVM memory occupation by MemoryMXBean:

![image](https://user-images.githubusercontent.com/5010047/170614454-84e7fe3e-71bb-4277-8c03-db0f461d6ccf.png)

Users are confused about why `HeapMemory+NonHeapMemory` is not equal to resident memory(top->res). It's still not clear what NonHeap means from `Java Doc`, I tried to understand the definition of NonHeap in source code, then found this bug...

Reply:
"non-heap" is any memory not part of the Java heap. But it is up to an implementation to decide what kinds of non-heap memory are represented by memory pools. The JavaDoc can only give possible examples not numerate an actual required set.